com.unity.netcode.gameobjects/Runtime/Core/SnapshotSystem.cs

```diff
@@ -459,16 +558,92 @@ internal void Despawn(SnapshotDespawnCommand command, NetworkObject networkObjec
             }
         }
 
-        // todo: entry-point for value updates
-        internal void Store(ulong networkObjectId, int behaviourIndex, int variableIndex, NetworkVariableBase networkVariable)
+        // entry-point for value updates
+        internal unsafe void Store(UpdateCommand command, NetworkVariableBase networkVariable)
         {
+            command.TickWritten = m_CurrentTick;
+            int commandPosition = -1;
+
+            List<ulong> targetClientIds = GetClientList();
+
+            if (targetClientIds.Count > 0)
+            {
+                // Look for an existing variable's position to update before adding a new entry
+                for (var i = 0; i < NumUpdates; i++)
+                {
+                    if (Updates[i].BehaviourIndex == command.BehaviourIndex &&
+                        Updates[i].NetworkObjectId == command.NetworkObjectId &&
+                        Updates[i].VariableIndex == command.VariableIndex)
+                    {
+                        commandPosition = i;
+                        break;
+                    }
+                }
+
+                if (commandPosition == -1)
+                {
+                    int index = -1;
+
+                    if (NumUpdates >= UpdatesBufferCount)
+                    {
+                        UpdatesBufferCount = UpdatesBufferCount * 2;
+                        Array.Resize(ref Updates, UpdatesBufferCount);
+                        Array.Resize(ref UpdatesMeta, UpdatesBufferCount);
+                    }
+
+                    commandPosition = NumUpdates;
+                    NumUpdates++;
+
+                    index = m_AvailableIndices[0];
+                    m_AvailableIndices[0] = m_AvailableIndices[m_NumAvailableIndices - 1];
+                    m_NumAvailableIndices--;
+
+                    UpdatesMeta[commandPosition].Index = index;
+                }
+                else
+                {
+                    // de-allocate previous buffer as a new one will be allocated
+                    MemoryStorage.Deallocate(UpdatesMeta[commandPosition].Index);
+                }
```

Comment:
this code block is a little weird to me.
on one hand, we do manual memory/array management but on the other hand, we use `IndexAllocator::Deallocate` as a helper/util?

Reply:
Discussed over Slack. Resolving for now.